src/Microsoft.DocAsCode.Build.OverwriteDocuments/OverwriteDocumentModelCreater.cs

```diff
@@ -3,13 +3,193 @@
+        }
+
+        private static void AppendNewObject(List<OPathSegment> OPathSegments, Block codeHeaderBlock, List<Block> propertyValue, Dictionary<string, object> contentsMetadata)
+        {
+            var currentObject = contentsMetadata;
+            for (int index = 0; index < OPathSegments.Count - 1; index++)
+            {
+                currentObject = FindOrCreateSegment(currentObject, codeHeaderBlock, OPathSegments[index]);
+            }
+
+            var lastSegment = OPathSegments.Last();
+            if (currentObject.ContainsKey(lastSegment.SegmentName))
+            {
+                if (currentObject[lastSegment.SegmentName] is List<Block>)
+                {
+                    // Duplicate
+                    Logger.LogWarning(
+                        $"There are two duplicate OPaths {OPathSegments.Select(o => o.OriginalSegmentString).Aggregate((a, b) => a + "/" + b)}, the previous one will be overwritten",
+                        line: codeHeaderBlock.Line.ToString(),
+                        code: WarningCodes.Overwrite.InvalidOPaths);
+                }
+                else
+                {
+                    throw new MarkdownFragmentsException(
+                        $"A({lastSegment.SegmentName}) is expected to be an dictionary with \"A/B\" or an array of dictionaries with \"A[c=d]/C\", however it is used as an array of Blocks in line {codeHeaderBlock.Line} with \"../A\" OPath syntax",
+                        codeHeaderBlock.Line);
+                }
+            }
+
+            currentObject[lastSegment.SegmentName] = propertyValue;
+        }
+
+        private static Dictionary<string, object> FindOrCreateSegment(Dictionary<string, object> currentObject, Block codeHeaderBlock, OPathSegment segment)
+        {
+            Dictionary<string, object> nextObject;
+            if (currentObject.ContainsKey(segment.SegmentName))
+            {
+                if (currentObject[segment.SegmentName] is List<Block>)
+                {
+                    throw new MarkdownFragmentsException(
+                        $"A({segment.SegmentName}) is expected to be an array of Blocks with \"../A\", however it is used as an dictionary or an array of dictionaries in line {codeHeaderBlock.Line} with \"A/B\" or \"A[c=d]/B\" OPath syntax",
+                        codeHeaderBlock.Line);
+                }
+
+                nextObject = currentObject[segment.SegmentName] as Dictionary<string, object>;
+                if (nextObject != null)
+                {
+                    if (!string.IsNullOrEmpty(segment.Key))
+                    {
+                        throw new MarkdownFragmentsException(
+                            $"A({segment.SegmentName}) is expected to be an object with \"A/B\", however it is used as an array in line {codeHeaderBlock.Line} with \"A[c=d]/C\" OPath syntax",
+                            codeHeaderBlock.Line);
```

Comment:
I think it's straightforward to revert the logic: first look up whether `segment.Key` is null, then cast to expected type. If cast failed, throw Exception. This can also reduce the call to `as`.